src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -86,6 +86,95 @@ internal WebApplicationBuilder(WebApplicationOptions options, Action<IHostBuilde
     }
 
+    internal WebApplicationBuilder(WebApplicationOptions options, bool slim, Action<IHostBuilder>? configureDefaults = null)
+    {
+        Debug.Assert(slim, "should only be called with slim: true");
+
+        var configuration = new ConfigurationManager();
+
+        // add the default host configuration sources, so they are picked up by the HostApplicationBuilder constructor.
+        // These won't be added by HostApplicationBuilder since DisableDefaults = true.
+        configuration.AddEnvironmentVariables(prefix: "DOTNET_");
+        if (options.Args is { Length: > 0 } args)
+        {
+            configuration.AddCommandLine(args);
+        }
+
+        _hostApplicationBuilder = new HostApplicationBuilder(new HostApplicationBuilderSettings
+        {
+            DisableDefaults = true,
+            Args = options.Args,
+            ApplicationName = options.ApplicationName,
+            EnvironmentName = options.EnvironmentName,
+            ContentRootPath = options.ContentRootPath,
+            Configuration = configuration,
+        });
+
+        // configure the ServiceProvider explicitly because DisableDefaults means HostApplicationBuilder won't.
+        var serviceProviderFactory = GetServiceProviderFactory(_hostApplicationBuilder);
+        _hostApplicationBuilder.ConfigureContainer(serviceProviderFactory);
+
+        // Set WebRootPath if necessary
+        if (options.WebRootPath is not null)
+        {
+            Configuration.AddInMemoryCollection(new[]
+            {
+                new KeyValuePair<string, string?>(WebHostDefaults.WebRootKey, options.WebRootPath),
+            });
+        }
+
+        // Run methods to configure web host defaults early to populate services
+        var bootstrapHostBuilder = new BootstrapHostBuilder(_hostApplicationBuilder);
+
+        // This is for testing purposes
+        configureDefaults?.Invoke(bootstrapHostBuilder);
+
+        bootstrapHostBuilder.ConfigureSlimWebHost(
+            webHostBuilder =>
+            {
+                AspNetCore.WebHost.UseKestrel(webHostBuilder);
+
+                webHostBuilder.Configure(ConfigureEmptyApplication);
+
+                webHostBuilder.UseSetting(WebHostDefaults.ApplicationKey, _hostApplicationBuilder.Environment.ApplicationName ?? "");
+                webHostBuilder.UseSetting(WebHostDefaults.PreventHostingStartupKey, Configuration[WebHostDefaults.PreventHostingStartupKey]);
+                webHostBuilder.UseSetting(WebHostDefaults.HostingStartupAssembliesKey, Configuration[WebHostDefaults.HostingStartupAssembliesKey]);
+                webHostBuilder.UseSetting(WebHostDefaults.HostingStartupExcludeAssembliesKey, Configuration[WebHostDefaults.HostingStartupExcludeAssembliesKey]);
+            },
+            options =>
+            {
+                // We've already applied "ASPNETCORE_" environment variables to hosting config
+                options.SuppressEnvironmentConfiguration = true;
```

Comment:
We have not already applied "ASPNETCORE_" environment variables to hosting config in this code path. Only "DOTNET_". We should do both. I bet this is why the macOS template test is failing to load the server certificate.

We had an [issue](https://github.com/dotnet/aspnetcore/issues/20479) a while back with the template tests using expired dev certs on macOS. That was [fixed](https://github.com/dotnet/aspnetcore/pull/20494) by using "ASPNETCORE_"-prefixed environment variables to set a default cert, so the tests shouldn't rely on the development cert being up to date.

https://github.com/dotnet/aspnetcore/blob/40dd2305cf26b3d3a5fa85473b66edf0fd14e40b/src/ProjectTemplates/Shared/AspNetProcess.cs#L91-L95

I'm guessing the dev cert is working as a fallback on the other OS's, but we still have that original issue on macOS when it falls back to the dev cert. @adityamandaleeka I wonder if this is related to our wider macOS dev cert issues.